core/vm/contracts_stateful.go

```diff
@@ -4,34 +4,40 @@
 package vm
 
 import (
+	"github.com/ava-labs/subnet-evm/precompile"
 	"github.com/ethereum/go-ethereum/common"
 )
 
-var (
-	AllowListPrecompileAddress = common.HexToAddress("0x0200000000000000000000000000000000000000")
-)
-
-// StatefulPrecompiledContract is the interface for executing a precompiled contract
-// This wraps the PrecompiledContracts native to Ethereum and allows adding in stateful
-// precompiled contracts to support native Avalanche asset transfers.
-type StatefulPrecompiledContract interface {
-	// Run executes a precompiled contract in the current state
-	// assumes that it has already been verified that [caller] can
-	// transfer [value].
-	Run(evm *EVM, caller ContractRef, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error)
-}
-
 // wrappedPrecompiledContract implements StatefulPrecompiledContract by wrapping stateless native precompiled contracts
 // in Ethereum.
 type wrappedPrecompiledContract struct {
 	p PrecompiledContract
 }
 
-func newWrappedPrecompiledContract(p PrecompiledContract) StatefulPrecompiledContract {
+// newWrappedPrecompiledContract returns a wrapped version of [PrecompiledContract] to be executed according to the StatefulPrecompiledContract
+// interface.
+func newWrappedPrecompiledContract(p PrecompiledContract) precompile.StatefulPrecompiledContract {
 	return &wrappedPrecompiledContract{p: p}
 }
 
 // Run implements the StatefulPrecompiledContract interface
-func (w *wrappedPrecompiledContract) Run(evm *EVM, caller ContractRef, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+func (w *wrappedPrecompiledContract) Run(accessibleState precompile.PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
 	return RunPrecompiledContract(w.p, input, suppliedGas)
 }
+
+// RequiredGas returns the amount of gas consumed by [w] when operating on [input].
+func (w *wrappedPrecompiledContract) RequiredGas(input []byte) uint64 {
+	return w.p.RequiredGas(input)
+}
+
+// RunStatefulPrecompiledContract confirms runs [precompile] with the specified parameters.
+func RunStatefulPrecompiledContract(precompile precompile.StatefulPrecompiledContract, accessibleState precompile.PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	// Confirm that there is sufficient gas to execute [precompile]
+	// Note: we perform this check here to so that we can return ErrOutOfGas if [precompile] will run out of gas.
+	requiredGas := precompile.RequiredGas(input)
+	if requiredGas > suppliedGas {
+		return nil, 0, ErrOutOfGas
+	}
+
```

Comment:
should we deduce `requiredGas` here instead of in the `Run`?

Reply:
Nah, I think it is good to do inside `Run` to be consistent